packages/backend/src/ws/WebSocketManager.ts

```diff
@@ -0,0 +1,111 @@
+export const AUTHORIZED = 'authorized';
+
+export class WebSocketManager implements IWebSocketManager {
+    private readonly connections: Map<string, Socket[]> = new Map();
+    private readonly web3Provider: ethers.providers.Web3Provider;
+    private readonly db: IDatabase;
+
+    /**
+     * @param {http.Server} httpServer - The HTTP server instance.
+     * @param {ethers.providers.Web3Provider} web3Provider - ethers web3provider instance.
+     * @param {IDatabase} db - The database instance.
+     */
+    constructor(
+        httpServer: http.Server,
+        web3Provider: ethers.providers.Web3Provider,
+        db: IDatabase,
+    ) {
+        //Establish Ws
+        const server = new Server(httpServer, {
+            cors: {
+                origin: '*',
+            },
+        });
+        this.web3Provider = web3Provider;
+        this.db = db;
+
+        //register listener
+        server.on('connection', (c: Socket) => {
+            this.addConnection(c);
+        });
+    }
+    /**
+     * Checks if a user is connected.
+     * @param {string} ensName - The ENS name of the user.
+     * @returns {boolean} - Returns true if the user is connected with at least one socket, false otherwise.
+     */
+    public async isConnected(ensName: string) {
+        const connections = this.connections.get(ensName);
+        return !!(connections && connections.length > 0);
+    }
+    /**
+     * Adds a new connection to the connections map.
+     * @private
+     * @param {Socket} connection - The socket connection instance.
+     */
+    private async addConnection(connection: Socket) {
+        try {
+            const { account, token } = connection.handshake.auth;
+
+            const ensName = normalizeEnsName(account.ensName);
+            //Use the already existing function cechkToken to check weather the token matches the provided ensName
+            const hasSession = await checkToken(
+                this.web3Provider,
+                this.db.getSession,
+                ensName,
+                token,
+            );
+            //retrive the session from the db
+            const session = await this.db.getSession(ensName);
+            //If the ensName has not a valid session we disconnect the socket
+            if (!hasSession || !session) {
+                console.log('connection refused for ', ensName);
+                connection.emit(UNAUTHORIZED);
+                connection.disconnect();
+            }
+            //Get the old connections and add the new one
+            const oldConnections = this.connections.get(ensName) || [];
+            this.connections.set(ensName, [...oldConnections, connection]);
+            //Send the authorized event
+            connection.emit(AUTHORIZED);
+            console.log('connection established for ', ensName);
+            //When the socket disconnects we wan't them no longer in our connections List
+            connection.on('disconnect', () => {
```

Comment:
did you check if anybody ever removes a connection if the client simply loses their internet connection? How does this happen? Is it after a timeout? If so, how long is that timeout?

Reply:
disconnect will be called if the connection has been interrupted regardless the reason
Can be either called programmatically by the client, or more likely terminated on the networking layer (the client has been closed)

In both cases the disconnected event will be invoked